src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -214,9 +179,67 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
             }
         }
 
-        private bool DetermineExtensionDataProperty(Dictionary<string, JsonPropertyInfo> cache)
+        private void InitializeConstructorParameters(ConstructorInfo constructorInfo)
+        {
+            ParameterInfo[] parameters = constructorInfo!.GetParameters();
+            Dictionary<string, JsonParameterInfo> parameterCache = CreateParameterCache(parameters.Length, Options);
+
+            foreach (ParameterInfo parameterInfo in parameters)
+            {
+                PropertyInfo? firstMatch = null;
+                bool isBound = false;
+
+                foreach (JsonPropertyInfo jsonPropertyInfo in PropertyCacheArray!)
+                {
+                    // This is not null because it is an actual
+                    // property on a type, not a "policy property".
+                    PropertyInfo propertyInfo = jsonPropertyInfo.PropertyInfo!;
+
+                    string camelCasePropName = JsonNamingPolicy.CamelCase.ConvertName(propertyInfo.Name);
+
+                    if (parameterInfo.Name == camelCasePropName &&
+                        parameterInfo.ParameterType == propertyInfo.PropertyType)
+                    {
+                        if (isBound)
+                        {
+                            Debug.Assert(firstMatch != null);
+
+                            // Multiple object properties cannot bind to the same
+                            // constructor parameter.
+                            ThrowHelper.ThrowInvalidOperationException_MultiplePropertiesBindToConstructorParameters(
+                                Type,
+                                parameterInfo,
+                                firstMatch,
+                                propertyInfo,
+                                constructorInfo);
+                        }
+
+                        JsonParameterInfo jsonParameterInfo = AddConstructorParameter(parameterInfo, jsonPropertyInfo, Options);
+
+                        // One object property cannot map to multiple constructor
+                        // parameters (ConvertName above can't return multiple strings).
+                        parameterCache.Add(jsonParameterInfo.NameAsString, jsonParameterInfo);
+
+                        isBound = true;
+                        firstMatch = propertyInfo;
+                    }
+                }
+            }
+
+            // It is invalid for the extension data property to bind with a constructor argument.
+            if (DataExtensionProperty != null &&
+                parameterCache.ContainsKey(DataExtensionProperty.NameAsString!))
+            {
+                ThrowHelper.ThrowInvalidOperationException_ExtensionDataCannotBindToCtorParam(DataExtensionProperty.PropertyInfo!, Type, constructorInfo);
+            }
+
+            ParameterCache = parameterCache;
+            ParameterCount = parameters.Length;
```

Comment:
I don't think this is thread-safe. `ParameterCache` may be used by another thread before `ParameterCount` is set. Do we actually need `ParameterCount` (can we just use `ParameterCache.Count`)?

Reply:
I believe it is.

`parameters` here is the parameter list for the constructor, which is constant. `ParameterCache` only contains parameters which map to properties, so the count values might be different.

When these count values are different, this means that not all parameters mapped to properties, so we throw an exception if deserialization is attempted: 

https://github.com/dotnet/runtime/blob/9202f23223310cf11a9f39b177e885cda699b72c/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs#L79-L82

I've added some comments.

Reply:
Are both values are set for the given ClassInfo instance before it is added to the ClassInfo cache? If so, then we should be OK w.r.t. threading.

Reply:
Yes, they are.